Modules/posixmodule.c

```diff
@@ -4214,7 +4214,204 @@ os_listdir_impl(PyObject *module, path_t *path)
+        Py_DECREF(result);
+        return NULL;
+    }
+
+    int err = 0;
+    Py_BEGIN_ALLOW_THREADS;
+    find = FindFirstVolumeW(buffer, 256);
+    if (find == INVALID_HANDLE_VALUE) {
+        err = GetLastError();
+    }
+    Py_END_ALLOW_THREADS;
+
+    while (!err) {
+        PyObject *s = PyUnicode_FromWideChar(buffer, -1);
+        if (!s || PyList_Append(result, s) < 0) {
+            Py_XDECREF(s);
+            Py_CLEAR(result);
+            break;
+        }
+        Py_DECREF(s);
+
+        Py_BEGIN_ALLOW_THREADS;
+        if (!FindNextVolumeW(find, buffer, 256)) {
+            err = GetLastError();
+        }
+        Py_END_ALLOW_THREADS;
+    }
+
+    if (find != INVALID_HANDLE_VALUE) {
+        Py_BEGIN_ALLOW_THREADS;
+        FindVolumeClose(find);
+        Py_END_ALLOW_THREADS;
+    }
+    if (err && err != ERROR_NO_MORE_FILES) {
+        PyErr_SetFromWindowsErr(err);
+        if (result) {
+            Py_CLEAR(result);
+        }
+    }
+    return result;
+}
+
+
+/*[clinic input]
+os.listmounts
+
+    volume: path_t
+
+Return a list containing mount points for a particular volume.
+
+'volume' should be a GUID path as returned from os.listvolumes.
+
+[clinic start generated code]*/
+
+static PyObject *
+os_listmounts_impl(PyObject *module, path_t *volume)
+/*[clinic end generated code: output=06da49679de4512e input=a8a27178e3f67845]*/
+{
+    DWORD buflen = 64;
+    wchar_t defaultBuffer[64];
+    LPWSTR buffer = defaultBuffer;
+    int err = 0;
+    if (PySys_Audit("os.listmounts", "O", volume->object) < 0) {
+        return NULL;
+    }
+    Py_BEGIN_ALLOW_THREADS;
+    if (!GetVolumePathNamesForVolumeNameW(volume->wide, buffer, buflen, &buflen)) {
+        err = GetLastError();
+        if (err == ERROR_MORE_DATA) {
+            Py_BLOCK_THREADS;
+            buffer = (wchar_t*)PyMem_Malloc(sizeof(wchar_t) * buflen);
+            Py_UNBLOCK_THREADS;
+            if (buffer) {
+                if (!GetVolumePathNamesForVolumeNameW(volume->wide, buffer,
+                                                      buflen, &buflen)) {
+                    err = GetLastError();
+                }
+            }
+        }
+    }
+    Py_END_ALLOW_THREADS;
+
+    PyObject *str = NULL;
+    PyObject *nullchar = NULL;
+    PyObject *r = NULL;
+
+    if (err) {
+        PyErr_SetFromWindowsErr(err);
+        goto exit;
+    }
+    if (!buffer) {
+        goto exit;
+    }
+    if (buflen < 2) {
+        r = PyList_New(0);
+        goto exit;
+    }
+
+    /* buflen includes two null terminators (one for the last string
+       and one for the array of strings */
+    str = PyUnicode_FromWideChar(buffer, buflen - 2);
+    nullchar = PyUnicode_FromStringAndSize("\0", 1);
+    if (str && nullchar) {
+        r = PyUnicode_Split(str, nullchar, buflen);
+    }
+exit:
+    if (buffer && buffer != defaultBuffer) {
+        PyMem_Free((void *)buffer);
+    }
+    Py_XDECREF(nullchar);
+    Py_XDECREF(str);
+    return r;
```

Comment:
Suggestions:

* increase the default buffer size to be large enough for a single mountpoint on a path that's up to `MAX_PATH - 1` characters, plus two terminating nulls
* use the `PY_ARRAY_SIZE` macro for the initial `buflen`
* use a while loop to avoid a race condition in case another thread adds a new mountpoint; improbable, but no point in not coding defensively
* avoid the need to use `Py_BLOCK_THREADS`; it's overly complicated
* set an exception via `PyErr_NoMemory()` if buffer allocation fails; `PyMem_Malloc()` and `PyMem_Realloc()` don't do this for us.
* don't rely on `err`, which mistakenly isn't reset to 0 here after handling `ERROR_MORE_DATA` successfully. it's simpler to just set the `OSError` exception and bail out after `GetVolumePathNamesForVolumeNameW()` fails.
* for `PyUnicode_Split()`, use the default value of -1 for `maxsplit` (no limit) instead of `buflen`. it's unclear why you pass the original buffer length (includes the two terminating nulls) as `maxsplit`. I guess it's an upper bound on the number of splits, but it's simpler to just use -1.

```suggestion
    wchar_t default_buffer[MAX_PATH + 1];
    DWORD buflen = Py_ARRAY_LENGTH(default_buffer);
    LPWSTR buffer = default_buffer;
    PyObject *str = NULL;
    PyObject *nullchar = NULL;
    PyObject *result = NULL;

    if (PySys_Audit("os.listmounts", "O", volume->object) < 0) {
        return NULL;
    }

    while (1) {
        BOOL success;
        Py_BEGIN_ALLOW_THREADS
        success = GetVolumePathNamesForVolumeNameW(volume->wide, buffer,
                                                   buflen, &buflen);
        Py_END_ALLOW_THREADS
        if (success) {
            break;
        }
        if (GetLastError() != ERROR_MORE_DATA) {
            PyErr_SetFromWindowsErr(GetLastError());
            goto exit;
        }
        void *temp = PyMem_Realloc(buffer == default_buffer ? NULL : buffer,
                                   sizeof(wchar_t) * buflen);
        if (!temp) {
            PyErr_NoMemory();
            goto exit;
        }
        buffer = temp;
    }

    if (buflen < 2) {
        result = PyList_New(0);
        goto exit;
    }

    // buflen includes two null terminators, one for the last string
    // and one for the array of strings.
    str = PyUnicode_FromWideChar(buffer, buflen - 2);
    nullchar = PyUnicode_FromStringAndSize("\0", 1);
    if (str && nullchar) {
        result = PyUnicode_Split(str, nullchar, -1);
    }

exit:
    if (buffer != default_buffer) {
        PyMem_Free(buffer);
    }
    Py_XDECREF(nullchar);
    Py_XDECREF(str);
    return result;
```

Reply:
`GetVolumePathNamesForVolumeNameW()` queries the mountpoint manager for the volume's registered pathnames, but it doesn't verify that the volume is actually mounted by a filesystem. A raw volume that hasn't been formatted can have registered pathnames. But they aren't mountpoints. 

I think `os.listmounts()` should first call `GetFileAttributesW(volume->wide)` to verify the path. Fail if the volume isn't formatted with a recognized filesystem.

```c
    DWORD attributes;
    Py_BEGIN_ALLOW_THREADS
    attributes = GetFileAttributesW(path->wide);
    Py_END_ALLOW_THREADS
    if (attributes == INVALID_FILE_ATTRIBUTES &&
        GetLastError() == ERROR_UNRECOGNIZED_VOLUME)
    {
        return PyErr_SetFromWindowsErr(ERROR_UNRECOGNIZED_VOLUME);
    }
```

Reply:
Ah, good call. I didn't even check what happens if you pass in invalid arguments...

Reply:
>  I didn't even check what happens if you pass in invalid arguments...

`GetVolumePathNamesForVolumeNameW()` requires `lpszVolumeName` to be the root path of a volume name of the form "\\\\?\\Volume{GUID}\\", else it fails with `ERROR_INVALID_PARAMETER`. The volume device has to exist and be known to the mountpoint manager, else it fails with `ERROR_FILE_NOT_FOUND`. However, neither the API nor the mountpoint manager requires that the volume is mounted by a recognized filesystem.

The volume path name has its uses even when the volume isn't mounted. For example, one can pass it to "format.com": `format "path\to\volume" /fs:exFAT /Q`. The format command accepts volume GUID names, but DOS path names are easier to work with and remember.